### ISSUE TITLE
help browser: shift+enter doesn't evaluate code

### ISSUE THREAD

**Issue by htor** [supercollider/supercollider#4542]
hitting shift+enter doesn't evaluate the current line under cursor in the help browser.

## Environment

* SuperCollider version: 3.10.2 develop [6b7e6ec58]
* Operating system: macos 10.13.6
* Other details (Qt version, audio driver, etc.): 5.13.0

## Steps to reproduce

open help browser and evaluate any line with shift+enter. 
it will do nothing

## Debugging

the shift+enter keybinding function gets called in js (`evalLine()`) but it does nothing.
i inserted a `console.log` into `function setUpWebChannel(port)` in `scdoc.js` but it never logs, so i suspect `setUpWebChannel(port)` is never called and therefore `evalLine()` in turn will do nothing since `window.IDE` will be `undefined`.

i see this output in my terminal: 

```
~/Programming/supercollider/build @develop $ ./Install/SuperCollider/SuperCollider.app/Contents/MacOS/SuperCollider
QAbstractSocket::connectToHost() called when already looking up or connecting/connected to "127.0.0.1"
QAbstractSocket::connectToHost() called when already looking up or connecting/connected to "127.0.0.1"
QAbstractSocket::connectToHost() called when already looking up or connecting/connected to "127.0.0.1"
QAbstractSocket::connectToHost() called when already looking up or connecting/connected to "127.0.0.1"
```

**Comment by brianlheim** [comment]
@htor please do not close issue tickets (including your own) that have merit without explanation. I understand you are upset, but that does not give you the right to behave this way.

**Comment by htor** [comment]
i really think bridging the gap between the help browser and the rest of the ide's functionality using a webchannel/socket is a good idea, although a tad unstable at the moment. if we can get this working we can get rid of all the js-based code evaluation logic (https://github.com/supercollider/supercollider/blob/develop/HelpSource/editor.js#L72-L167), which is not as much tested and works less ideal than the ide's. we've already discussed the possibilites of sharing settings, themes so the help browser will be visually more integrated, so that's also nice. any insights , @snappizz  @brianlheim? cheers

**Comment by elgiano** [comment]
I looked a bit at this, and there is this onPageLoad() function which as far as I understand nobody ever calls:
https://github.com/supercollider/supercollider/blob/7c48b5166c259fb37c417341e6d8bec3ba28b134/editors/sc-ide/widgets/help_browser.cpp#L116

I've tried to connect it as a slot to the "page loaded" signal from qwebengineview... and then everything works. Any reason why it should stay disconnected? I admit I haven't looked at all into websockets yet.. so maybe @snappizz could have some ideas about it?

Disclosure: my goal in working on this is that if shift+enter works, then ctrl+enter could evaluate code blocks :)

**Comment by snappizz** [comment]
Unfortunately I'm not interested in maintaining help browser code eval. I think it's better to have users copy-paste code to SCIDE files rather than force ourselves to indefinitely maintain two editors, one Qt and the other in the browser.

**Comment by brianlheim** [comment]
i am still interested in this, even if @snappizz is not :) @elgiano that part of the codebase was rewritten somewhat recently and it's very possible mistakes were made in the process. if that seems to solve the issue then feel free to submit a PR!

**Comment by snappizz** [comment]
Reading my previous comment and realizing it came off as unnecessarily grumpy. I'm sorry about that. It's just not a feature I use personally.

**Comment by brianlheim** [comment]
no worries! it's okay if you're not interested, i feel the same way about other aspects of supercollider myself :) just want to make sure this fix doesn't get lost if there are people that will benefit from it.

**Comment by htor** [comment]
i don't get how "i don't use this feature myself" is a valid excuse for not dealing with bugs or features. yes, you don't get payed, blah blah, open source, etc etc. but you choose to be a maintainer, so you have to deal with issues you "don't like to maintain" too. the fact is users are having a bad experience using the supercollider documentation because it's broken and therefore this is a bad thing. either deal with the problem or remove the feature entirely. don't dismiss it as not important because you have a different workflow. a broken feature is just going to generate much more noise and you're going to have to deal with it sooner or later.

**Comment by htor** [comment]
https://github.com/elgiano/rfcs/blob/helpbrowser/rfcs/0009-helpbrowser-code-evaluation.md
there's now a fix + rfc (related to my comment above) in the making from a new contributor. i really hope you welcome this contributor with open arms, even though the help browser is not interesting to you "personally".

**Comment by dvzrv** [comment]
> i don't get how "i don't use this feature myself" is a valid excuse for not dealing with bugs or features. yes, you don't get payed, blah blah, open source, etc etc. but you choose to be a maintainer, so you have to deal with issues you "don't like to maintain" too. the fact is users are having a bad experience using the supercollider documentation because it's broken and therefore this is a bad thing. either deal with the problem or remove the feature entirely. don't dismiss it as not important because you have a different workflow. a broken feature is just going to generate much more noise and you're going to have to deal with it sooner or later.

@htor please don't make this a personal tirade on someone. SuperCollider is a free software project for better or worse. A handful of developers are actively working on the gazillion issues that this cross-platform monster (and I mean that in the best possible way) has. Discouraging them by personal attacks will not get you anywhere.

That all being said: I understand why users want code execution capabilities straight from help browser and I also understand why certain other users can't (due to licensing issues) or won't have qtwebengine in their help browser. Both use-cases need to remain valid (I guess) and that makes decision making on this topic not necessarily straight forward.

> https://github.com/elgiano/rfcs/blob/helpbrowser/rfcs/0009-helpbrowser-code-evaluation.md
> there's now a fix + rfc (related to my comment above) in the making from a new contributor. i really hope you welcome this new contributor with open arms, even though the help browser is not interesting to you "personally".

This is great news and makes me wonder why you had to post the comment before that?

**Comment by htor** [comment]
> This is great news and makes me wonder why you had to post the comment before that?

consider it a contribution. free software, right?

**Comment by dvzrv** [comment]
> > This is great news and makes me wonder why you had to post the comment before that?
> 
> consider it a contribution. free software, right?

I would like to remind you of the [code of conduct](https://github.com/supercollider/supercollider/blob/develop/CODE_OF_CONDUCT.md) that we strive to communicate by.
Please take a step back and think about your way of communication here.

**Comment by jamshark70** [comment]
@htor I guess what I would add here is two-fold: 1/ the devs are hearing your point about how important help browser code execution is to you and other users (and there's concrete progress on it); 2/ when you're replying to someone who is apologizing for overstating a case, it may be better to accept the apology rather than snip back.

That is, the two are separate: the point about user need is legit, but that doesn't justify the tone of those replies.

The legit point is being heard and worked on. That's a good thing.

**Comment by brianlheim** [comment]
@htor consider this your final warning. personal attacks against maintainers of this project are completely unacceptable, and violate our code of conduct. please apologize to Nathan or i will have to ask you to leave.

**Comment by htor** [comment]
i'll see myself out

**Comment by brianlheim** [comment]
i'm sorry to hear that.

i've blocked htor from participating in the github organization, and marked the unfortunate digression here as off-topic. my apologies especially to @snappizz and @dvzrv and to anyone else reading this.

**Comment by brianlheim** [comment]
fixed in #4883, thanks @elgiano !
